en: update docs about shard DDL lock (#269)
pingcap/docs-dm: en/feature-manually-handling-sharding-ddl-locks.md

{
    "result": true,                                        # The result of the unlocking operation.
    "msg": "",                                             # The additional message for the failure to unlock the lock.
}
```

## Supported scenarios

Currently, the `unlock-ddl-lock` command only supports handling sharding DDL locks in the following two abnormal scenarios.

### Scenario 1: Some MySQL sources are removed

#### The reason for the abnormal lock

> Before `DM-master` tries to automatically unlock the sharding DDL lock, all the MySQL sources need to receive the sharding DDL events (for details, see [shard merge principles](feature-shard-merge.md#principles)). If the sharding DDL event is already in the replication process, and some MySQL sources have been removed and are not to be reloaded (these MySQL sources have been removed according to the application demand), then the sharding DDL lock cannot be automatically replicated and unlocked because not all the DM-workers can receive the DDL event.

Reviewer: ```suggestion
Before `DM-master` tries to automatically unlock the sharding DDL lock, all the MySQL sources need to receive the sharding DDL events (for details, see [shard merge principles](feature-shard-merge-pessimistic.md#principles)). If the sharding DDL event is already in the replication process, and some MySQL sources have been removed and are not to be reloaded (these MySQL sources have been removed according to the application demand), then the sharding DDL lock cannot be automatically replicated and unlocked because not all the DM-workers can receive the DDL event.
```

#### Impact

After you have manually unlocked the lock by using `unlock-ddl-lock`, if you don't deal with the offline MySQL sources included in the task configuration information, the lock might still be unable to be replicated automatically when the next sharding DDL event is received.

Therefore, after you have manually unlocked the DDL lock, you should perform the following operations:

1. Use `stop-task` to stop the running tasks.
2. Update the task configuration file, and remove the related information of the offline MySQL source from the configuration file.
3. Use `start-task` and the new task configuration file to restart the task.

> **Note:**
>
> After you run `unlock-ddl-lock`, if the MySQL source that went offline becomes online again and tries to replicate the data of the sharded tables, a match error between the data and the downstream table structure might occur.

> ### Scenario 2: Some DM-workers stop abnormally or  network failure occurs during the DDL unlocking process

Reviewer 2: ```suggestion
### Scenario 2: Some DM-workers stop abnormally or network failure occurs during the DDL unlocking process
```

#### Manual solution

Suppose that now we have the same upstream and downstream table structures and the same demand for merging tables and replication as in the manual solution of [Some MySQL sources are removed](#scenario-1-some-mysql-sources-are-removed).

When `DM-master` automatically executes the unlocking process, the owner （`mysql-replica-01`）successfully executes the DDL and continues the replication process, but in the process of requesting the non-owner (`mysql-replica-02`) to skip the DDL operation, the checkpoint fails to update after skipping the DDL because the corresponding DM worker was restarted.

After `DM-worker-2` restarts, it will try to replicate the waiting DDL lock before it restarted. At this point, a new lock will be created on `DM-master`, and the DM-worker will become the owner of the lock (other DM-workers have executed/skipped the DDL by now and are continuing the replication process).

> After the data migration subtask corresponding to `mysql-replica-02` restores, a new lock will be created on the DM-master, but other MySQL sources are already executed or skipped DDL operations and are performing subsequent replication.

Reviewer 2: blank line before this 
```suggestion

After the data migration subtask corresponding to `mysql-replica-02` restores, a new lock will be created on the DM-master, but other MySQL sources are already executed or skipped DDL operations and are performing subsequent replication.
```

3. Use `start-task` and the new task configuration file to restart the task.

> **Note:**
>
> After you run `unlock-ddl-lock`, if the MySQL source that went offline becomes online again and tries to replicate the data of the sharded tables, a match error between the data and the downstream table structure might occur.

### Scenario 2: Some DM-workers stop abnormally or the network failure occurs during the DDL unlocking process

#### The reason for the abnormal lock

After `DM-master` receives the DDL events of all DM-workers, automatically running `unlock DDL lock` mainly include the following steps:

1. Ask the owner of the lock to execute the DDL and update the checkpoints of corresponding sharded tables.
2. Remove the DDL lock information stored on `DM-master` after the owner successfully executes the DDL.
3. Ask all other non-owners to skip the DDL and update the checkpoints of corresponding sharded tables after the owner successfully executes the DDL.
4. DM-master removes the corresponding DDL lock information after all owner or non-owner operations are successful.

> Currently, the above unlocking process is not atomic. If the non-owner skips the DDL operation successfully, the DM-worker where it is located stops abnormally or a network exception occurs with the downstream TiDB, which will cause a failure to update the checkpoint successfully.

Reviewer: ```suggestion
Currently, the above unlocking process is not atomic. If the non-owner skips the DDL operation successfully, the DM-worker where the non-owner is located stops abnormally or a network anomaly occurs with the downstream TiDB, which can cause the checkpoint updating to fail.
```

> When the non-owner corresponding MySQL source restores data migration, it tries to request the DM-master to re-coordinate the DDL that has been coordinated before the exception occurs and never receive the corresponding DDL from other MySQL sources, which will cause the DDL operation to automatically unlock the corresponding lock.

Reviewer: ```suggestion
When the MySQL source corresponding to the non-owner restores data migration, the non-owner tries to request the DM-master to re-coordinate the DDL operation that has been coordinated before the exception occurs and will never receives the corresponding DDL operation from other MySQL sources. This can cause the DDL operation to automatically unlock the corresponding lock.
```

#### Manual solution

Suppose that now we have the same upstream and downstream table structures and the same demand for merging tables and replication as in the manual solution of [Some MySQL sources are removed](#scenario-1-some-mysql-sources-are-removed).

> When `DM-master` automatically executes the unlocking process, the owner (`mysql-replica-01`) successfully executes the DDL and continues the replication process, but in the process of requesting the non-owner (`mysql-replica-02`) to skip the DDL operation, the checkpoint fails to update after skipping the DDL because the corresponding DM worker was restarted.

Reviewer: ```suggestion
When `DM-master` automatically executes the unlocking process, the owner (`mysql-replica-01`) successfully executes the DDL and continues the replication process. However, in the process of requesting the non-owner (`mysql-replica-02`) to skip the DDL operation, the checkpoint fails to update after the DM-worker skips the DDL operation because the corresponding DM-worker was restarted.
```

> After `DM-worker-2` restarts, it will try to replicate the waiting DDL lock before it restarted. At this point, a new lock will be created on `DM-master`, and the DM-worker will become the owner of the lock (other DM-workers have executed/skipped the DDL by now and are continuing the replication process).
> 
> After the data migration subtask corresponding to `mysql-replica-02` restores, a new lock will be created on the DM-master, but other MySQL sources are already executed or skipped DDL operations and are performing subsequent replication.

Reviewer: ```suggestion
After the data migration subtask corresponding to `mysql-replica-02` restores, a new lock is created on the DM-master, but other MySQL sources have executed or skipped DDL operations and are performing subsequent replication.
```

2. Use `unlock-ddl-lock` to ask `DM-master` to unlock the lock.

>     - Try to execute the DDL to the downstream again during the unlocking process (the owner before restarting has executed the DDL to the downstream), so as to make sure that the DDL can be executed multiple times.

Reviewer: ```suggestion
    - During the unlocking process, the owner tries to execute the DDL operation to the downstream again (the original owner before restarting has executed the DDL operation to the downstream once). Make sure that the DDL operation can be executed multiple times.
```
